src/main/java/net/neoforged/neoforge/event/entity/player/PlayerRespawnEvent.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event.entity.player;
+
+import javax.annotation.Nullable;
+import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.LevelReader;
+import net.minecraft.world.phys.Vec3;
+import net.neoforged.fml.LogicalSide;
+import net.neoforged.neoforge.common.NeoForge;
+
+/**
+ * Fired by {@link net.minecraft.server.players.PlayerList#respawn(ServerPlayer, boolean)} when the server respawns a player.
+ * This may be used to change the {@link ServerLevel} the player respawns in, as well as their respawn position.
+ * This event is fired after {@link net.minecraft.world.level.block.state.BlockState#getRespawnPosition(EntityType, LevelReader, BlockPos, float, LivingEntity)} is called.
+ * <p>
+ * This event is fired on the {@linkplain NeoForge#EVENT_BUS main NeoForge event bus},
+ * only on the {@linkplain LogicalSide#SERVER logical server}.
+ */
+public class PlayerRespawnEvent extends PlayerEvent {
+    private ServerLevel respawnLevel;
+    private @Nullable Vec3 respawnPosition;
+    private float respawnAngle;
+    private final ServerLevel originalRespawnLevel;
+    private final @Nullable Vec3 originalRespawnPosition;
+    private final float originalRespawnAngle;
+    private final boolean fromEndFight;
+    private boolean changePlayerSpawnPosition = false;
+
+    public PlayerRespawnEvent(ServerPlayer player, ServerLevel respawnLevel, float respawnAngle, @Nullable Vec3 respawnPosition, boolean fromEndFight) {
+        super(player);
+        this.respawnLevel = respawnLevel;
+        this.respawnPosition = respawnPosition;
+        this.respawnAngle = respawnAngle;
+        this.originalRespawnLevel = respawnLevel;
+        this.originalRespawnPosition = respawnPosition;
+        this.originalRespawnAngle = respawnAngle;
+        this.fromEndFight = fromEndFight;
+    }
+
+    /**
+     * The level the player will respawn into
+     */
+    public ServerLevel getRespawnLevel() {
+        return respawnLevel;
+    }
+
+    public void setRespawnLevel(ServerLevel respawnLevel) {
+        this.respawnLevel = respawnLevel;
+    }
```

Comment:
An overload should be provided to set the respawn level via `ResourceKey` and perform that lookup rather than relying on the user to do it (since the resource key -> level lookup is typically annoying).